### PR TITLE
feat(eslint-plugin): add extension rule `init-declarations`

### PR DESCRIPTION
**Pull request by anikethsaha**
fixes https://github.com/typescript-eslint/typescript-eslint/issues/1811

- init-declarations
- added tests for declaring of typed variables
- docs

### CODE REVIEW

**Comment by anikethsaha** [comment]
@bradzacher `messageIds` is simply extending `string` and that is causing error for test file's messaged of invalid cases. Any workaround.? Is it ok to change the type in `ruletester` to `string` instead of extending it .

similar for `errors.type`

File: packages/eslint-plugin/docs/rules/init-declarations.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# require or disallow initialization in variable declarations (`init-declarations`)
+
+In Typescript, variables can be assigned during declaration, or at any point afterwards using an assignment statement. For example, in the following code, `foo` is initialized during declaration, while `bar` is initialized later.
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/init-declarations.md]: A good example of how to write rule docs for an extension rule is here
https://github.com/typescript-eslint/typescript-eslint/blob/7d963fd846935acd91b7b0cd31c56a70a2b994d1/packages/eslint-plugin/docs/rules/no-dupe-class-members.md

Essentially - we don't want to copy+paste the docs, as they can get out of sync with ESLint's docs.
So instead we want to just mention what changes we make (like "This rule extends the base `eslint/init-declarations` rule. It adds support for TypeScript's variable declaration syntax.")

File: packages/eslint-plugin/typings/eslint-rules.d.ts [hunk 2]
```diff
@@ -543,3 +543,21 @@ declare module 'eslint/lib/rules/no-extra-semi' {
   >;
   export = rule;
 }
+
+declare module 'eslint/lib/rulesinit-declarations' {
+  import { TSESLint, TSESTree } from '@typescript-eslint/experimental-utils';
+
+  const rule: TSESLint.RuleModule<
+    never,
```
Comment (bradzacher) [inline packages/eslint-plugin/typings/eslint-rules.d.ts]: you can use the message IDs from the base rule here
https://github.com/eslint/eslint/blob/0c20bc068e608869981a10711bba88ffde1539d8/lib/rules/init-declarations.js#L89-L92

File: packages/eslint-plugin/docs/rules/init-declarations.md (hunk 1 above)
Comment (anikethsaha) [reply packages/eslint-plugin/docs/rules/init-declarations.md]: cool, I will add a reference

File: packages/eslint-plugin/typings/eslint-rules.d.ts (hunk 2 above)
Comment (anikethsaha) [reply packages/eslint-plugin/typings/eslint-rules.d.ts]: @bradzacher I didnt get this. no other rules are having `messageIds` as type in the module.

Comment (bradzacher) [reply packages/eslint-plugin/typings/eslint-rules.d.ts]: The ones with rules that support messageIds do have them declared, i.e.

https://github.com/typescript-eslint/typescript-eslint/blob/1e29e69b289d61107a7de67592beae331ba50222/packages/eslint-plugin/typings/eslint-rules.d.ts#L7-L14

File: packages/eslint-plugin/docs/rules/init-declarations.md (hunk 1 above)
Comment (anikethsaha) [reply packages/eslint-plugin/docs/rules/init-declarations.md]: done

File: packages/eslint-plugin/typings/eslint-rules.d.ts (hunk 2 above)
Comment (anikethsaha) [reply packages/eslint-plugin/typings/eslint-rules.d.ts]: done

File: packages/eslint-plugin/typings/eslint-rules.d.ts [hunk 3]
```diff
@@ -543,3 +543,21 @@ declare module 'eslint/lib/rules/no-extra-semi' {
   >;
   export = rule;
 }
+
+declare module 'eslint/lib/rulesinit-declarations' {
```
Comment (bradzacher) [inline packages/eslint-plugin/typings/eslint-rules.d.ts]:
```suggestion
declare module 'eslint/lib/rules/init-declarations' {
```

Comment (anikethsaha) [reply packages/eslint-plugin/typings/eslint-rules.d.ts]: oops...sorry

**Comment by bradzacher** [comment]
make sure you validate your types locally and fix the errors.

https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#validating-your-changes

File: packages/eslint-plugin/src/rules/init-declarations.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: ['always'],
+  create(context) {
+    const rules = baseRule.create(context);
+    const mode = context.options[0] || 'always';
+
+    return {
+      'VariableDeclaration:exit'(
+        node: TSESTree.VariableDeclaration | TSESTree.Node,
```
Comment (anikethsaha) [inline packages/eslint-plugin/src/rules/init-declarations.ts]: @bradzacher for both 

`TSESTree.VariableDeclaration | TSESTree.Node`

and

`TSESTree.VariableDeclaration | ts.Node`

The same errors are coming about the `declare` not found. `VariableDeclaration` has that but still, I dont know why its not able to locate it.

Comment (anikethsaha) [reply packages/eslint-plugin/src/rules/init-declarations.ts]: Should I create a custom type for this node ? it seems `VariableDeclaration` and `Node` doesn't have common type fields like `parent` and `declare`

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/init-declarations.ts]: `TSESTree.VariableDeclaration | TSESTree.Node` is the incorrect type.
`TSESTree.Node` is the union of all the node types, so `TSESTree.VariableDeclaration | TSESTree.Node === TSESTree.Node`.

`parent` is always optional due to how the parser works - this is intentional.
```suggestion
        node: TSESTree.VariableDeclaration,
```

File: packages/eslint-plugin/typings/eslint-rules.d.ts [hunk 5]
```diff
@@ -556,7 +556,9 @@ declare module 'eslint/lib/rules/init-declarations' {
       }?,
     ],
     {
-      'VariableDeclaration:exit'(node: TSESTree.VariableDeclaration): void;
+      'VariableDeclaration:exit'(
+        node: TSESTree.VariableDeclaration | TSESTree.Node,
```
Comment (bradzacher) [inline packages/eslint-plugin/typings/eslint-rules.d.ts]:
```suggestion
        node: TSESTree.VariableDeclaration,
```

File: packages/eslint-plugin/src/rules/init-declarations.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      'VariableDeclaration:exit'(
+        node: TSESTree.VariableDeclaration,
+      ): void {
+        if (mode === 'always') {
+          if (node?.declare) {
+            return;
+          }
+          if (
+            node?.parent?.type === AST_NODE_TYPES.TSModuleBlock &&
+            node?.parent?.parent?.declare
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/init-declarations.ts]: Unlike in plain javascript where you can make assumptions and write working code that does few checks, when writing typescript you have to be explicit.

In this instance, we as engineers can logically say that a TSModuleBlock can only have a parent that is TSModuleDeclaration.
However the types are setup more generally such that a TSModuleBlock has a parent that is of type Node - i.e. the types represent that any node type could be its parent.

Now for safety - typescript does not let you access a property unless it exists on every single type in a union. This is done for safety and simplicity.

So you have to do one of two things:
1) assert that `declare` exists on `node.parent.parent`.
2) assert that `node.parent.parent` is of type TSModuleDeclaration.

(1) is a lot less desired than (2), as it does not assert on the expected structure of the AST.

Comment (anikethsaha) [reply packages/eslint-plugin/src/rules/init-declarations.ts]: this is giving error free check.

```ts
   node.parent = <TSESTree.TSModuleBlock>node?.parent;
   node.parent.parent = <TSESTree.TSModuleDeclaration>(
            node?.parent?.parent
          );
```

Comment (anikethsaha) [reply packages/eslint-plugin/src/rules/init-declarations.ts]: should I add this ?

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/init-declarations.ts]: that's one way to do it, but it is a type-only check. This means that if for some weird reason the node isn't a TSModuleDeclaration, so it's now made the code unsafe.

A _better_ solution would be a runtime assertion, which typescript can use to narrow the type for you:
```suggestion
            node.parent?.type === AST_NODE_TYPES.TSModuleBlock &&
            node.parent?.parent.type === AST_NODE_TYPES. TSModuleDeclaration &&
            node.parent?.parent?.declare
```

**Comment by anikethsaha** [comment]
@bradzacher looks like there is some auth issue, I cant push to this branch after reading 

image

**Comment by bradzacher** [comment]
Make sure your local is up to date with the remote before pushing, or try force pushing if you want to overwrite the remote.

**Comment by anikethsaha** [comment]
I did rebase and after that its coming.

**Comment by anikethsaha** [comment]
@bradzacher I did add some more tests but the diff coverage is still the same as before adding them

**Comment by bradzacher** [comment]
https://codecov.io/gh/typescript-eslint/typescript-eslint/pull/1814/diff

The lines marked with yellow are branches that haven't been tested.

I.e. looking at your code + tests, you've got code that handles `module` declarations, but no tests for it.

**Comment by anikethsaha** [comment]
image

@bradzacher I don't know how to particularly test those lines. I did add some test cases to cover those conditions.

**Comment by bradzacher** [comment]
> I don't know how to particularly test those lines.

In codecov, yellow means that one branch of a conditional check has been tested, but not the other branch.

So looking at the code in question:
```ts
if (
  node.parent?.type === AST_NODE_TYPES.TSModuleBlock &&
  node.parent?.parent?.type === AST_NODE_TYPES.TSModuleDeclaration &&
  node.parent?.parent?.declare
) {
  return;
}
```

What is this if statement doing? What is it guarding against?
Compare it with your tests - which of the branches are you testing, and which are you not?

(this is not me trying to be snarky here - I'm trying to drop a hint without giving you the answer 😄).

**Comment by anikethsaha** [comment]
> (this is not me trying to be snarky here - I'm trying to drop a hint without giving you the answer 😄)

:sweat_smile:


I will give another try 👍

**Comment by anikethsaha** [comment]
@bradzacher 
I did add some tests for those if statements in the latest commits 
https://github.com/typescript-eslint/typescript-eslint/pull/1814/commits/ed71f47f58ae02ec7dff6276faf113af909d5823

https://github.com/typescript-eslint/typescript-eslint/pull/1814/commits/de1544f898f68589583a613b90076289ef5659f7

https://github.com/typescript-eslint/typescript-eslint/pull/1814/commits/07950f6a8be75a49268f52f7af912dc8067b9fbc

anything missing, a hint would be great

**Comment by bradzacher** [comment]
Those test are sufficient - the reason codecov is reporting lower is because of the optional chain.
The parent is never nullable, so there's no way to test that branch - just ignore it. Your coverage is sufficient.

The branch target isn't a hard merge blocker. It's a guideline to help ensure people do enough tests and help me review that the tests cover the cases.

**Comment by anikethsaha** [comment]
>  the reason codecov is reporting lower is because of the optional chain.

I doubted that. idk even codecov supports the optional chaining or not

> The branch target isn't a hard merge blocker. It's a guideline to help ensure people do enough tests and help me review that the tests cover the cases.

cool, 👍

**Pull request merged**